### PR TITLE
Use of write_multiscale_labels for image

### PR DESCRIPTION
**Pull request by will-moore**
Testing `write_multiscale_labels()` added in https://github.com/ome/ome-zarr-py/pull/178/

### CODE REVIEW

**Comment by sbesson** [comment]
This feature has now been released as part of `ome-zarr 0.4.0` .

**Review by sbesson — CHANGES_REQUESTED** [review]
Tested using the same set of samples as the ones generated in #116 

To compare the output, I used the following minimal script which should compare arrays and flag difference in attributes

```python
import zarr
import sys

z1 = zarr.open(sys.argv[1])
z2 = zarr.open(sys.argv[2])

if z1.attrs != z2.attrs:
  print("Different top-level attributes")

assert len(z1) == len(z2)


for a in z1.arrays():
  key = a[0]
  assert (z1[key][:] == z2[key][:]).all()

if z1['labels'].attrs != z2['labels'].attrs:
  print("Different labels attributes")

for g in z1['labels'].groups():
  label_name = g[0]
  if z1['labels'][label_name].attrs != z2['labels'][label_name].attrs:
    print("Different image-label attributes")
  assert len(z1['labels'][label_name]) == len(z2['labels'][label_name])
  for a in z1['labels'][label_name].arrays():
    key = a[0]
    assert (z1['labels'][label_name][key][:] == z2['labels'][label_name][key][:]).all()
```

The execution of this script reveals no difference in the binary data content when comparing the arrays but a few mismatches in the attributes 

```
(conversion) [sbesson@pilot-zarr1-dev omero-cli-zarr_116]$ for i in *.zarr; do echo $i && python diff.py $i ../omero-cli-zarr_107/$i; done
10501752.zarr
Different top-level attributes
Different image-label attributes
4496763.zarr
Different top-level attributes
Different image-label attributes
5514375.zarr
Different top-level attributes
Different image-label attributes
Different image-label attributes
6001247.zarr
Different top-level attributes
Different image-label attributes
```

The top-level attribute difference comes from the version of the creation tool and is expected

```diff
(conversion) [sbesson@pilot-zarr1-dev omero-cli-zarr_116]$ for i in *.zarr; do diff $i/.zattrs ../omero-cli-zarr_107/$i/.zattrs; done
4c4
<         "version": "0.3.1.dev33+gf305941"
---
>         "version": "0.3.1.dev40+gcd84f02"
4c4
<         "version": "0.3.1.dev33+gf305941"
---
>         "version": "0.3.1.dev40+gcd84f02"
4c4
<         "version": "0.3.1.dev33+gf305941"
---
>         "version": "0.3.1.dev40+gcd84f02"
4c4
<         "version": "0.3.1.dev33+gf305941"
---
>         "version": "0.3.1.dev40+gcd84f02"
```

For the attribute of the `image-label` layer, the new code adds the version to the image-label specification but the multiscales name attribute seems to be lost and should be restored

```diff
(conversion) [sbesson@pilot-zarr1-dev omero-cli-zarr_116]$ find * -mindepth 2 -iname .zattrs -exec diff ../omero-cli-zarr_107/{} {} \; 
23991c23991,23992
<         }
---
>         },
>         "version": "0.4"
24076d24076
<             "name": "0",
6029c6029,6030
<         }
---
>         },
>         "version": "0.4"
6152d6152
<             "name": "0",
609c609,610
<         }
---
>         },
>         "version": "0.4"
700d700
<             "name": "Cell",
609c609,610
<         }
---
>         },
>         "version": "0.4"
700d700
<             "name": "Chromosomes",
681c681,682
<         }
---
>         },
>         "version": "0.4"
764d764
<             "name": "0",
```

I will also have a go at testing a plate labels export

**Comment by sbesson** [comment]
Unfortunately, my test of plate labels export failes using sample plates from `idr0001` as the masks derived from the polygon ROIs seem to be overlapping

```
(conversion) [sbesson@pilot-zarr1-dev omero-cli-zarr_107]$ omero zarr polygons Plate:2669
Using session for public@idr.openmicroscopy.org:4064. Idle timeout: 10 min. Current group: Public
Found 0 mask shapes in 0 ROIs
Found 19 mask shapes in 19 ROIs
Unique dimensions: {'T': {None}, 'C': {None}, 'Z': {None}}
source_image 2669.zarr/D/9/1
Ignoring dimensions {'Z', 'C', 'T'}
Traceback (most recent call last):
  File "/home/sbesson/miniconda3/envs/conversion/bin/omero", line 10, in <module>
    sys.exit(main())
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero/main.py", line 125, in main
    rv = omero.cli.argv()
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero/cli.py", line 1784, in argv
    cli.invoke(args[1:])
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero/cli.py", line 1222, in invoke
    stop = self.onecmd(line, previous_args)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero/cli.py", line 1299, in onecmd
    self.execute(line, previous_args)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero/cli.py", line 1381, in execute
    args.func(args)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero_zarr/cli.py", line 102, in _wrapper
    return func(self, *args, **kwargs)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero_zarr/cli.py", line 298, in polygons
    plate_shapes_to_zarr(plate, ["Polygon"], args)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero_zarr/masks.py", line 89, in plate_shapes_to_zarr
    saver.save(list(masks.values()), args.label_name)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero_zarr/masks.py", line 301, in save
    labels, fill_colors, properties = self.masks_to_labels(
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero_zarr/masks.py", line 523, in masks_to_labels
    raise Exception(
Exception: Mask 413187 overlaps with existing labels
(conversion) [sbesson@pilot-zarr1-dev omero-cli-zarr_107]$ omero zarr polygons Plate:2551
Using session for public@idr.openmicroscopy.org:4064. Idle timeout: 10 min. Current group: Public
Found 0 mask shapes in 0 ROIs
Found 0 mask shapes in 0 ROIs
Found 0 mask shapes in 0 ROIs
Found 0 mask shapes in 0 ROIs
Found 0 mask shapes in 0 ROIs
Found 0 mask shapes in 0 ROIs
Found 65 mask shapes in 65 ROIs
Unique dimensions: {'T': {None}, 'C': {None}, 'Z': {None}}
source_image 2551.zarr/D/5/0
Ignoring dimensions {'C', 'Z', 'T'}
Traceback (most recent call last):
  File "/home/sbesson/miniconda3/envs/conversion/bin/omero", line 10, in <module>
    sys.exit(main())
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero/main.py", line 125, in main
    rv = omero.cli.argv()
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero/cli.py", line 1784, in argv
    cli.invoke(args[1:])
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero/cli.py", line 1222, in invoke
    stop = self.onecmd(line, previous_args)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero/cli.py", line 1299, in onecmd
    self.execute(line, previous_args)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero/cli.py", line 1381, in execute
    args.func(args)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero_zarr/cli.py", line 102, in _wrapper
    return func(self, *args, **kwargs)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero_zarr/cli.py", line 298, in polygons
    plate_shapes_to_zarr(plate, ["Polygon"], args)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero_zarr/masks.py", line 89, in plate_shapes_to_zarr
    saver.save(list(masks.values()), args.label_name)
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero_zarr/masks.py", line 301, in save
    labels, fill_colors, properties = self.masks_to_labels(
  File "/home/sbesson/miniconda3/envs/conversion/lib/python3.9/site-packages/omero_zarr/masks.py", line 523, in masks_to_labels
    raise Exception(
Exception: Mask 632718 overlaps with existing labels
```

The export of a single image with non overlapping polygons works as expected


```
(conversion) [sbesson@pilot-zarr1-dev omero-cli-zarr_107]$ omero zarr export Image:1229945
Using session for public@idr.openmicroscopy.org:4064. Idle timeout: 10 min. Current group: Public
Exporting to 1229945.zarr (0.4)
Finished.
(conversion) [sbesson@pilot-zarr1-dev omero-cli-zarr_107]$ omero zarr polygons Image:1229945
Using session for public@idr.openmicroscopy.org:4064. Idle timeout: 10 min. Current group: Public
Export Polygons on Image: JL_120731_S6A [Well F-8; Field #1]
Found 3 mask shapes in 3 ROIs
Unique dimensions: {'T': {None}, 'C': {None}, 'Z': {None}}
source_image 1229945.zarr
Ignoring dimensions {'Z', 'T', 'C'}
```

Barring the plate label export workflow, the only remaining issue is the loss of the `version` in the `image-label` dictionary.

**Comment by sbesson** [comment]
https://github.com/ome/ome-zarr-py/pull/206 proposes to address the regression mentioned in https://github.com/ome/omero-cli-zarr/pull/107#pullrequestreview-1007701234 by updating the underlying API to write the version metadata.

File: setup.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def get_long_description() -> str:
     author="The Open Microscopy Team",
     author_email="",
     python_requires=">=3",
-    install_requires=["omero-py>=5.6.0", "ome-zarr>=0.3.0"],
+    install_requires=["omero-py>=5.6.0", "ome-zarr>=0.4.0"],
```
Comment (sbesson) [inline setup.py:38]:
```suggestion
    install_requires=["omero-py>=5.6.0", "ome-zarr>=0.5.0"],
```

which includes the API addition allowing to preserve the version metadata in the `image-label` attributes

**Review by sbesson — CHANGES_REQUESTED** [review]

**Comment by will-moore** [comment]
@sbesson Thanks for releasing ome-zarr-py 0.5.0 and bumping the requirements here.
This good to merge now?

**Comment by sbesson** [comment]
I am retesting the PR with the data regeneration and validation process described above. Once this is successful, yes I'd propose to merge and release `omero-cli-zarr`. Should I also regenerate and upload some (all?) of the sample OME-NGFF datasets and add them to idr.github.io/ome-ngff-samples/ as examples of label images with 8-bit and/or 16-bit types?

**Comment by will-moore** [comment]
I would like to update as many of the sample OME-NGFF datasets as possible, since that allows viewing the labels on the web (zarr.js). I guess it *might* be useful to keep a sample as 64-bit, in case someone wants to test with it, but in general I'd like to replace existing labels, since having both is confusing / redundant. Happy to help with regeneration/update etc.

**Review by sbesson — APPROVED** [review]
Retested the conversion workflow and confirmed that the arrays are identical and the only metadata difference are 1- the creator version 2- the addition of image names

```
(conversion) [sbesson@pilot-zarr1-dev omero-cli-zarr_107]$ for i in *.zarr; do echo $i && python diff.py $i ../omero-cli-zarr_116/$i; done
10501752.zarr
Different top-level attributes
Different image-label attributes
4496763.zarr
Different top-level attributes
Different image-label attributes
5514375.zarr
Different top-level attributes
Different image-label attributes
Different image-label attributes
6001247.zarr
Different top-level attributes
Different image-label attributes
(conversion) [sbesson@pilot-zarr1-dev omero-cli-zarr_107]$ for i in *.zarr; do diff $i/.zattrs ../omero-cli-zarr_116/$i/.zattrs; done
4c4
<         "version": "0.3.1.dev41+g5c73169"
---
>         "version": "0.3.1.dev33+gf305941"
4c4
<         "version": "0.3.1.dev41+g5c73169"
---
>         "version": "0.3.1.dev33+gf305941"
4c4
<         "version": "0.3.1.dev41+g5c73169"
---
>         "version": "0.3.1.dev33+gf305941"
4c4
<         "version": "0.3.1.dev41+g5c73169"
---
>         "version": "0.3.1.dev33+gf305941"
(conversion) [sbesson@pilot-zarr1-dev omero-cli-zarr_107]$ find * -mindepth 2 -iname .zattrs -exec diff ../omero-cli-zarr_116/{} {} \;
24076a24077
>             "name": "0",
6152a6153
>             "name": "0",
700a701
>             "name": "Cell",
700a701
>             "name": "Chromosomes",
764a765
>             "name": "0",
```

and confirmed that the `dtype` of the labels is still int8 or int16 as per #107 

```
(conversion) [sbesson@pilot-zarr1-dev omero-cli-zarr_107]$ find * -mindepth 3 -iname .zarray -exec grep dtype {} \;
    "dtype": "<i2",
    "dtype": "<i2",
    "dtype": "<i2",
    "dtype": "<i2",
    "dtype": "<i2",
    "dtype": "<i2",
    "dtype": "<i2",
    "dtype": "<i2",
    "dtype": "<i2",
    "dtype": "<i2",
    "dtype": "<i2",
    "dtype": "<i2",
    "dtype": "|i1",
    "dtype": "|i1",
    "dtype": "|i1",
    "dtype": "|i1",
    "dtype": "|i1",
    "dtype": "|i1",
    "dtype": "|i1",
    "dtype": "|i1",
    "dtype": "|i1",
    "dtype": "|i1",
    "dtype": "|i1",
    "dtype": "|i1",
```

**Pull request merged**